waku/protocol/v2/waku_store.nim

```diff
@@ -173,11 +173,89 @@ proc encode*(rpc: HistoryRPC): ProtoBuffer =
+    dir = pinfo.direction
+
+  if pageSize == 0: # pageSize being zero indicates that no pagination is required
+    return (list, pinfo)
+
+  if list.len == 0: # no pagination is needed for an empty list
+    return (list, PagingInfo(pageSize: 0, cursor:pinfo.cursor, direction: pinfo.direction))
+
+  var msgList:seq[IndexedWakuMessage]= list # makes a copy of the list
+  msgList.sort(indexedWakuMessageComparison) # sorts msgList based on the custom comparison proc indexedWakuMessageComparison
+
+  var initQuery : bool = false
+  if cursor == Index(): 
+    initQuery = true # an empty cursor means it is an intial query
+    case dir
+      of PagingDirection.FORWARD: 
+        cursor = list[0].index # perform paging from the begining of the list
+      of PagingDirection.BACKWARD: 
+        cursor = list[list.len - 1].index # perform paging from the end of the list
+      
+  
+  var foundIndex = msgList.findIndex(cursor) 
+  if foundIndex == -1: # the cursor is not valid
+    return (@[], PagingInfo(pageSize: 0, cursor:pinfo.cursor, direction: pinfo.direction))
+
+  var retrievedPageSize, s, e: int
+  var newCursor: Index # to be returned as part of the new paging info
+  case dir
+    of PagingDirection.FORWARD: # forward pagination
+      let remainingMessages= msgList.len - foundIndex - 1
+      retrievedPageSize = min( int(pageSize), MaxPageSize).min(remainingMessages)  # the number of queried messages cannot exceed the MaxPageSize and the total remaining messages i.e., msgList.len-foundIndex
+      if initQuery : foundIndex = foundIndex - 1
+      s = foundIndex + 1  # non inclusive
+      e = foundIndex + retrievedPageSize 
+      newCursor = msgList[e].index # the new cursor points to the end of the page
+    of PagingDirection.BACKWARD: # backward pagination
+      let remainingMessages=foundIndex
+      retrievedPageSize = min( int(pageSize), MaxPageSize).min(remainingMessages) # the number of queried messages cannot exceed the MaxPageSize and the total remaining messages i.e., foundIndex-0
+      if initQuery : foundIndex = foundIndex + 1
+      s = foundIndex - retrievedPageSize 
+      e = foundIndex - 1
+      newCursor = msgList[s].index # the new cursor points to the begining of the page
+
+  # retrieve the messages
+  for i in s..e:
+    result[0].add(msgList[i])
+
+  result[1] = PagingInfo(pageSize : uint64(retrievedPageSize), cursor : newCursor, direction : pinfo.direction)
+
+
+proc paginateWithoutIndex( list: seq[IndexedWakuMessage], pinfo: PagingInfo): (seq[WakuMessage], PagingInfo) =
+  ## takes list, and perfomrs paging based on pinfo 
+  ## returns the page i.e, a sequence of WakuMessage and the new paging info to be used for the next paging request  
+  var (indexedData, updatedPagingInfo)= paginateWithIndex(list,pinfo)
+  for indexedMsg in indexedData:
+    result[0].add(indexedMsg.msg)
+  result[1]=updatedPagingInfo
+
 proc findMessages(w: WakuStore, query: HistoryQuery): HistoryResponse =
```

Comment:
Should this be a public function?

Reply:
This is not my implementation, @decanus knows better